File: 5-two-pointers/2-two-pointers-technique/main.java
// Two Pointers Technique - Java Example

public class TwoPointersTechnique {

    public static int removeDuplicates(int[] arr) {
        if (arr.length == 0) return 0;
        int writeIndex = 1;
        for (int readIndex = 1; readIndex < arr.length; readIndex++) {
            if (arr[readIndex] != arr[readIndex - 1]) {
                arr[writeIndex] = arr[readIndex];
                writeIndex++;
            }
        }
        return writeIndex;
    }

    public static void main(String[] args) {
        int[] arr = {1,1,2,2,3,4,4};
        int newLength = removeDuplicates(arr);

        System.out.print("Array without duplicates: ");
        for (int i = 0; i < newLength; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }
}

File: 5-two-pointers/2-two-pointers-technique/main.py
# Two Pointers Technique - Python Example

def remove_duplicates(arr):
    if not arr:
        return 0
    write_index = 1
    for read_index in range(1, len(arr)):
        if arr[read_index] != arr[read_index - 1]:
            arr[write_index] = arr[read_index]
            write_index += 1
    return write_index  # New length of array without duplicates

arr = [1,1,2,2,3,4,4]
new_length = remove_duplicates(arr)
print("Array without duplicates:", arr[:new_length])
